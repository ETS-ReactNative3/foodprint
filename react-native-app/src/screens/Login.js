import React, { useState, useRef } from 'react';
import { View, ScrollView } from 'react-native';
import { Button, Input, Text } from 'react-native-elements';
import Password from '../components/Password';
import Email from '../components/Email';
import { AuthContext } from '../store/Auth';

const Login = ({ navigation }) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const passwordRef = useRef(null);
  const [isPressed, setIsPressed] = useState(false);


  const { signIn } = React.useContext(AuthContext);

  const loginHandler = async () => {
    setIsPressed(true);
    await signIn({ email, password });
    setIsPressed(false);
  };

  return (
    <ScrollView contentContainerStyle={{ alignItems: 'center' }} style={{ width: '100%', height: '100%' }}>
      <View style={{ alignItems: 'center', justifyContent: 'center', marginVertical: 50 }}>
        <Text h1>
          FoodPrint
        </Text>
      </View>
      <View style={{ width: '80%' }}>
        <View>
          <Email nextFieldRef={passwordRef} setEmail={setEmail} email={email} />
          <Password ref={passwordRef} submitHandler={loginHandler} setPassword={setPassword} password={password} />
        </View>
      </View>
      <View style={{ width: '80%' }}>
        <Button
          disabled={isPressed}
          buttonStyle={{ backgroundColor: 'green', marginVertical: 100 }}
          titleStyle={{ fontSize: 24 }}
          title="Login"
          onPress={loginHandler}
        />
      </View>
    </ScrollView>
  );
};

export default Login;
