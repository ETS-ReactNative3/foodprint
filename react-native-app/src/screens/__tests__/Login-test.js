import React from 'react';

import renderer from 'react-test-renderer';
import { fireEvent, render, wait } from '@testing-library/react-native';
import Login from '../Login';
import AuthContext from '../../context/AuthContext';

jest.mock('@apollo/react-hooks');

// Check current component against its snapshot
test('Login matches previous snapshot', () => {
  const authentication = {
    signIn: () => { },
  };

  const tree = renderer.create(<AuthContext.Provider value={authentication}>
    <Login />
  </AuthContext.Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});


test('Login button calls signIn function from context if email and password is valid', async () => {
  const signIn = jest.fn();

  const authentication = {
    signIn,
  };

  const utils = renderLogin(authentication);

  const emailInput = utils.getByPlaceholderText('Email');
  const famousLecturerAtImperial = 'WillKnottenbelt@NotARealEmail.co.uk';
  fireEvent.changeText(emailInput, famousLecturerAtImperial);

  const passwordInput = utils.getByPlaceholderText('Password');
  const famousLecturersPassword = 'b10ckch@1n';
  fireEvent.changeText(passwordInput, famousLecturersPassword);

  const loginButton = utils.getByText('Login');
  fireEvent.press(loginButton);
  await wait(() => expect(utils.queryByTestId('loginButton')));
  expect(signIn.mock.calls.length).toBe(1);
});

test('Login button shows error if email is not valid', async () => {
  const utils = renderLogin({});

  const emailInput = utils.getByPlaceholderText('Email');
  const famousLecturerAtImperial = 'WillKnottenbelt@NotARealEmail';
  fireEvent.changeText(emailInput, famousLecturerAtImperial);

  const passwordInput = utils.getByPlaceholderText('Password');
  const famousLecturersPassword = 'b10ckch@1n';
  fireEvent.changeText(passwordInput, famousLecturersPassword);

  const loginButton = utils.getByText('Login');
  fireEvent.press(loginButton);
  await wait(() => expect(utils.queryByTestId('loginButton')));
  // TODO add expect error here (look for string)
});

// HELPERS

// Re-use this function to render Signup, as it uses useContext. This 'mocks' that out.
function renderLogin(authentication) {
  return render(
    <AuthContext.Provider value={authentication}>
      <Login />
    </AuthContext.Provider>
  );
}

