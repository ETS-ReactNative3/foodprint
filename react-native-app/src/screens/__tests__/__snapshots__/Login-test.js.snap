// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches previous snapshot 1`] = `
<ScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
    }
  }
  keyboardShouldPersistTaps="handled"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "justifyContent": "center",
        "marginVertical": 66.5,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/logoGreen.png",
        }
      }
      style={
        Object {
          "height": 200,
          "resizeMode": "contain",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "width": "80%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "marginBottom": 15,
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCompleteType="email"
            autoCorrect={false}
            editable={true}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Email"
            rejectResponderTermination={true}
            returnKeyType="next"
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginBottom": 15,
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCompleteType="password"
            autoCorrect={false}
            editable={true}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Password"
            rejectResponderTermination={true}
            returnKeyType="join"
            secureTextEntry={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "width": "80%",
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 3,
          "marginVertical": 16,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.3}
        disabled={false}
        onPress={[Function]}
        testID="loginButton"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "green",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 37.5,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            LOGIN
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</ScrollView>
`;
